src/EditorFeatures/Core/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,211 @@
+
+                symbol = symbol.GetOriginalUnreducedDefinition();
+                var project = document.Project;
+
+                // Get the symbol back from the originating workspace
+
+                var symbolMappingService = document.Project.Solution.Workspace.Services.GetService<ISymbolMappingService>();
+                var mappingResult = await symbolMappingService.MapSymbolAsync(document, symbol, cancellationToken).ConfigureAwait(false);
+                if (mappingResult != null)
+                {
+                    symbol = mappingResult.Symbol;
+                    project = mappingResult.Project;
+                }
+
+                var solution = project.Solution;
+                var sourceDefinition = await SymbolFinder.FindSourceDefinitionAsync(mappingResult.Symbol, mappingResult.Project.Solution, cancellationToken).ConfigureAwait(false);
+                if (sourceDefinition != null)
+                {
+                    var originatingProject = solution.GetProject(sourceDefinition.ContainingAssembly, cancellationToken);
+                    project = originatingProject ?? project;
+                }
+
+                // Three choices here:
+                // 1. Another language (like XAML) will take over via ISymbolNavigationService
+                // 2. There are locations in source, so we'll use those
+                // 3. There are no locations from source, so we'll try to generate a metadata as source file and use that
+                var results = new ArrayBuilder<CodeDefinitionWindowLocation>();
+                string filePath = null;
+                int lineNumber = 0;
+                int charOffset = 0;
+                var symbolNavigationService = solution.Workspace.Services.GetService<ISymbolNavigationService>();
+                var wouldNavigate = false;
+                await Task.Factory.StartNew(
+                    () => wouldNavigate = symbolNavigationService.WouldNavigateToSymbol(symbol, solution, out filePath, out lineNumber, out charOffset),
+                    cancellationToken,
+                    TaskCreationOptions.None,
+                    foregroundTaskScheduler).ConfigureAwait(false);
+
+                if (wouldNavigate)
+                {
+                    results.Add(new CodeDefinitionWindowLocation(symbol.ToDisplayString(), filePath, lineNumber, charOffset));
+                }
+                else
+                {
+                    var sourceLocations = symbol.Locations.Where(l => l.IsInSource).ToList();
+                    if (sourceLocations.Any())
+                    {
+                        foreach (var declaration in sourceLocations)
+                        {
+                            var declarationLocation = declaration.GetLineSpan();
+                            results.Add(new CodeDefinitionWindowLocation(symbol.ToDisplayString(), declarationLocation.Path, declarationLocation.Span.Start.Line, declarationLocation.Span.Start.Character));
```

Comment:
Should your CodeDefinitionWindowLocation type just use something else for line/column? It seems every place you construct one you have to deconstruct some structure.